Some GitHub plugin fixes

pmiossec opened the pull request:
## Proposed changes

-  ViewPullRequestsForm: 
    - better handle error on one remote (form is now still usable when a remote is on error)
    - fix selection of first remote found with at least a pull request
- CreatePullRequestForm: 
    - better handle error on one remote (form is now still usable when a remote is on error)
    - CreatePullRequestForm: increase width of dropdown 

## Screenshots <!-- Remove this section if PR does not change UI -->

### Before

Unable to do one easily but red arrow in 'after' screenshot show width increase to display well long branch name

### After

![image](https://user-images.githubusercontent.com/460196/114780658-fca0d580-9d77-11eb-8340-787c075887a7.png)

- ViewPullRequestsForm:

![image](https://user-images.githubusercontent.com/460196/115632346-cc24e280-a307-11eb-8b3a-d12797eba8d6.png)

- CreatePullRequestForm: 

![image](https://user-images.githubusercontent.com/460196/115632436-ff677180-a307-11eb-9a39-df9b3b7d4377.png)


## Test methodology <!-- How did you ensure quality? -->

- Manual


## Test environment(s) <!-- Remove any that don't apply -->

- Git Extensions 33.33.33
- Build 47bb7083c58cafdb7d180c425a88389b116b6f44
- Git 2.28.0.windows.1 (recommended: 2.30.0 or later)
- Microsoft Windows NT 10.0.19042.0
- .NET Framework 4.8.4341.0
- DPI 192dpi (200% scaling)

pmiossec commented:
@DamitSe I don't know exactly the reason of your error describe in comment https://github.com/gitextensions/gitextensions/issues/8850#issuecomment-819678693 but this PR could perhaps fix your issue and let you use the form...

DamitSe commented:
> @DamitSe I don't know exactly the reason of your error describe in comment [#8850 (comment)](https://github.com/gitextensions/gitextensions/issues/8850#issuecomment-819678693) but this PR could perhaps fix your issue and let you use the form...

I opened a new issue for this #9076; the issue was that I hadn't authorized the PAT but the dialog box didn't tell me that.

ivangrek commented on GitUI/CommandsDialogs/RepoHosting/CreatePullRequestForm.cs:
```diff
@@ -132,31 +132,37 @@ private void PopulateBranchesComboAndEnableCreateButton(IHostedRemote remote, Co
 
-                            comboBox.Items.Add(branches[i].Name);
-                        }
+                            if (branches.Count > 0)
```
`branches.Any()`

drewnoakes replied:
Why? This will be slower.

ivangrek replied:
It is more readable, more flexible, and uses a named method to denote an operation.
But I agree, this is subjective.

ivangrek replied:
Maybe extension method `Any()` for list will be better. 😄

RussKie replied:
Having worked on the .NET SDK for the past few years I've learnt to avoid Linq.

drewnoakes replied:
To be clear, my original objection wasn't really about performance. I guess it's what you're used to. I'm perfectly used to seeing a count check. My brain understands immediately, and also doesn't worry about allocations or overhead. If I see `Any` I worry about those things slightly, and worry about multiple enumeration and so forth more. But it's a matter of style.

I find these kinds of style things in PR reviews can provide a micro-barrier to contribution. In general I think projects at this kind of scale should use analyzers to enforce these kinds of preferences to save everyone's time and reduce friction. Of course that sometimes means writing an analyzer :)

ivangrek commented on GitUI/CommandsDialogs/RepoHosting/ViewPullRequestsForm.cs:
```diff
@@ -73,22 +73,27 @@ private void ViewPullRequestsForm_Load(object sender, EventArgs e)
+                        {
+                            hostedRemote.GetHostedRepository(); // We do this now because we want to do it in the async part.
+                        }
+                        catch (Exception)
```
```
catch
{}
```

RussKie replied:
Why?

drewnoakes replied:
In theory in IL it's possible to throw an arbitrary object. Catching `Exception` and an unspecified catch are technically different. I have never seen it make a difference in practice.

This is another example of something where an analyzer should step in to enfore this kind of thing according to the project's preferences IMO.

ivangrek commented on GitUI/CommandsDialogs/RepoHosting/ViewPullRequestsForm.cs:
```diff
@@ -99,7 +104,17 @@ private void _selectedOwner_SelectedIndexChanged(object sender, EventArgs e)
+            {
+                hostedRepo = hostedRemote?.GetHostedRepository();
+            }
+            catch (Exception)
```
```
catch
{}
```

RussKie commented on GitUI/CommandsDialogs/RepoHosting/CreatePullRequestForm.cs:
```diff
@@ -132,31 +132,37 @@ private void PopulateBranchesComboAndEnableCreateButton(IHostedRemote remote, Co
+                        catch (Exception)
                         {
-                            comboBox.SelectedIndex = selectItem;
                         }
```
This isn't good. @mstv and I have been working to remove these silent catch blocks. 
The error must be bubbled up to the bug reporter, and explain the user what can be done to resolve it, if possible.

RussKie commented on GitUI/CommandsDialogs/RepoHosting/ViewPullRequestsForm.cs:
```diff
@@ -73,22 +73,27 @@ private void ViewPullRequestsForm_Load(object sender, EventArgs e)
+                        {
+                            hostedRemote.GetHostedRepository(); // We do this now because we want to do it in the async part.
+                        }
+                        catch (Exception)
```
ditto

RussKie replied:
I think here we'd want to let the user know that we couldn't reach certain remotes. Otherwise if the user won't see some of the expected remotes they may not be able to immediately reconcile it.
Perhaps show a task dialog-based message:
```
We couldn't fetch the following repositories:
  - https://bla/bla/bla.git
  - https://foo/foo/foo.git
```

pmiossec replied:
I did something like that first but it will be more difficult to achieve: the error is displayed twice because the method is called from 2 places and I was not able to understand if that's normal and what will be the consequences to remove the first call :(

RussKie commented:
Ok, leave it like this but please add a comment

On Wed, 21 Apr. 2021, 11:05 pm Philippe Miossec, ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In GitUI/CommandsDialogs/RepoHosting/ViewPullRequestsForm.cs
> <https://github.com/gitextensions/gitextensions/pull/9075#discussion_r617513540>
> :
>
> >                      {
> -                        el.GetHostedRepository(); // We do this now because we want to do it in the async part.
> +                        try
> +                        {
> +                            hostedRemote.GetHostedRepository(); // We do this now because we want to do it in the async part.
> +                        }
> +                        catch (Exception)
>
> I did something like that first but it will be more difficult to achieve:
> the error is displayed twice because the method is called from 2 places and
> I was not able to understand if that's normal and what will be the
> consequences to remove the first call :(
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/gitextensions/gitextensions/pull/9075#discussion_r617513540>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABBTEXW6N6675ORKTZLK6LTTJ3ERBANCNFSM426FWDJQ>
> .
>

pmiossec commented:
> Ok, leave it like this but please add a comment
> 

I also think it is better to display a message so I can introduce a boolean to avoid the display of message when initializing on form load. What do you think?

pmiossec commented:
Added display of errors to the user. Added screenshots in the first post.

RussKie commented on GitUI/CommandsDialogs/RepoHosting/CreatePullRequestForm.cs:
```diff
@@ -132,31 +141,38 @@ private void PopulateBranchesComboAndEnableCreateButton(IHostedRemote remote, Co
+                        catch (Exception ex)
                         {
-                            comboBox.SelectedIndex = selectItem;
+                            MessageBox.Show(string.Format(_strRemoteInError.Text, remote.DisplayData) + Environment.NewLine + Environment.NewLine + ex.Message, TranslatedStrings.Error, MessageBoxButtons.OK, MessageBoxIcon.Warning);
```
A better error presentation:

```
[Error]

Remote: <xyz/uvw>
Error: Failed to load branches.
<...>
```

```
[Error]

Remote: <xyz/uvw>
Error: Remote ignored.
<...>
```

pmiossec replied:
@RussKie Is that what you have in mind?

![image](https://user-images.githubusercontent.com/460196/117469195-bd6c3b80-af55-11eb-92c3-1cfa8e2baff2.png)

![image](https://user-images.githubusercontent.com/460196/117469302-db39a080-af55-11eb-8bf6-12fccec89502.png)

RussKie replied:
Yes, though it looks a little awkward. 

I think it may look better if tweak it a little more:

* use TaskDialog (preferably from System.Windows.Forms namespace)
* Reformat as:
```
[ Remote ignored ] (<-- caption)
---------------------
The GitHub authentica..... (<-- main instruction aka Heading)

Remote: xx/yy/zz (Text)
```

Do you think you could try that?

pmiossec replied:
I did something like that using `Microsoft.WindowsAPICodePack.Dialogs`.
Indeed it was looking better.
I rebased and covert it to `System.Windows.Forms.TaskDialog`.
Unfortunately, the plugins no more load in .net 5 version.

So even if it compiles, we are unable to test it and we will have to wait for plugins loading....

RussKie commented:
Yep.
I'm working on reinstating all things back. I'm under a massive pump at
work atm, so have to prioritise things.

On Thu, 13 May 2021, 8:56 am Philippe Miossec, ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In GitUI/CommandsDialogs/RepoHosting/CreatePullRequestForm.cs
> <https://github.com/gitextensions/gitextensions/pull/9075#discussion_r631452206>
> :
>
> >                          {
> -                            comboBox.SelectedIndex = selectItem;
> +                            MessageBox.Show(string.Format(_strRemoteInError.Text, remote.DisplayData) + Environment.NewLine + Environment.NewLine + ex.Message, TranslatedStrings.Error, MessageBoxButtons.OK, MessageBoxIcon.Warning);
>
> I did something like that using Microsoft.WindowsAPICodePack.Dialogs.
> Indeed it was looking better.
> I rebased and covert it to System.Windows.Forms.TaskDialog.
> Unfortunately, the plugins no more load in .net 5 version.
>
> So even if it compiles, we are unable to test it and we will have to wait
> for plugins loading....
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/gitextensions/gitextensions/pull/9075#discussion_r631452206>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABBTEXUZBRFRRQTHOBRCDTDTNMBRFANCNFSM426FWDJQ>
> .
>

RussKie commented:
Let's merge this, and revisit when plugins are working again

pmiossec commented:
> Let's merge this, and revisit when plugins are working again

Yes, it won't change the face of the world.
I have rebased and squashed some commits...

RussKie commented:
Merge at will

On Fri, 21 May 2021, 8:20 am Philippe Miossec, ***@***.***>
wrote:

> Let's merge this, and revisit when plugins are working again
>
> Yes, it won't change the face of the world.
> I have rebased and squashed some commits...
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/gitextensions/gitextensions/pull/9075#issuecomment-845518257>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABBTEXTZGVMEVR22HODHLTLTOWDJDANCNFSM426FWDJQ>
> .
>
